### ISSUE TITLE
Getting bad TCN performance

### ISSUE THREAD

**Issue by bjourne** [philipperemy/keras-tcn#145]
Hi. I'm using your TCN for sequence prediction and I'm getting quite bad results in comparison to an LSTM. Maybe I'm doing something wrong or maybe there is a bug?

I'm comparing the units using:

    from tcn import TCN
    from tensorflow.keras import *
    from tensorflow.keras.layers import *
    from tensorflow.keras.optimizers import *
    import numpy as np

    SEQ_LEN = 64
    text = open('shakespeare.txt').read()
    ch2ix = {ch : i for i, ch in enumerate(sorted(set(text)))}
    X = np.array([[ch2ix[ch] for ch in text[i - SEQ_LEN:i]]
                  for i in range(SEQ_LEN, len(text))])
    Y = np.array([ch2ix[text[i]] for i in range(SEQ_LEN, len(text))])

    model = Sequential([
        Embedding(len(ch2ix), 32, input_shape = (SEQ_LEN,)),
        LSTM(128),
        # TCN(nb_filters = 64,
        #     kernel_size = 2,
        #     nb_stacks = 2,
        #     dilations = [1, 2, 4, 8]),
        Dense(len(ch2ix), activation = 'softmax')])
    opt = RMSprop(learning_rate = 0.005)
    model.compile(
        optimizer = opt,
        loss = 'sparse_categorical_crossentropy',
        metrics = ['sparse_categorical_accuracy'])
    model.fit(X, Y, epochs = 10, verbose = 1, batch_size = 32)

The number of parameters for the models are 113k for the TCN and 93k for the LSTM. After one epoch which takes the LSTM 216 seconds the loss is 1.81. For the TCN it is 2.31 and takes 339 seconds. "shakespeare.txt" is 1.1m text file containing Shakespeare plays.

**Comment by philipperemy** [comment]
@bjourne can you run it a bit more than just one epoch?

**Comment by rjpg** [comment]
Also in my cases studies WaveNet is better ...

**Comment by philipperemy** [comment]
I am revisiting this issue.

On the smaller version of the shakespeare.txt dataset, I could get:

- TCN: 48.7% accuracy on the test set.
- LSTM: 49.5% accuracy on the test set.

So the gap observed before is not present anymore.

**Comment by philipperemy** [comment]
I'll close this as the recent PRs have fixed most of the problems described in this PR. The LSTM still outperforms the TCN here (on the shakespeare.txt - 92kb) because the problem is more suitable for LSTM (char[t+1] heavily depends on char[t] and char[t-1] and not much 200 steps behind).
